src/cpp/src/continuous_batching_impl.cpp

```diff
@@ -513,17 +466,15 @@ ContinuousBatchingPipeline::ContinuousBatchingImpl::generate(const std::vector<o
             drop_requests(); // remove all requests from pipeline state in case of exception
             thrown_exception = std::current_exception();
         }
-        has_active_requests = has_non_finished_requests();
-        cv.notify_one();
+        stream_tokens(streamer_ptr, generation);
         if (thrown_exception) {
-            throw thrown_exception;
+            streamer_ptr->end();
+            std::rethrow_exception(thrown_exception);
```

Comment:
It's getting hard to follow. It seems that the intention is to call `end()` because the streamer would hang otherwise, `drop_requests()` to avoid memory leak and to `stream_tokens()` (I didn't figure out what for, can you explain?) in case of an error. RAII could be a cleaner solution. Introduce a class that would call all that in its destructor

Reply:
maybe `end()` should be called in `catch()` section? to avoid confusion

Reply:
I suppose if we have an exception, new token is not produced.

Reply:
@as-suvorov We can place end() inside catch section of course. But anyway we should rethrow exception

Reply:
`end()` moved to catch block